docs/userguide/exceptions.rst

```diff
@@ -110,3 +112,57 @@ as they are unaffected by task C's failure.
       (F)           (F)           (!F)
 
       time ----->
+
+
+Retry handlers
+--------------
+
+The basic parsl retry mechanism keeps a count of the number of times a task
+has been (re)tried, and will continue retrying that task until the configured
+retry limit is reached.
+
+Retry handlers generalize this to allow more expressive retry handling:
+parsl keeps a retry cost for a task, and the task will be retried until the
+configured retry limit is reached. Instead of the cost being 1 for each
+failure, user-supplied code can examine the failure and compute a custom
+cost.
+
+This allows user knowledge about failures to influence the retry mechanism:
+an exception which is almost definitely a non-recoverable failure (for example,
+due to bad parameters) can be given a high retry cost (so that it will not
+be retried many times, or at all), and exceptions which are likely to be
+transient (for example, where a worker node has died) can be given a low
+retry cost so they will be retried many times.
+
+A retry handler can be specified in the parsl configuration like this:
+
+
+.. code-block:: python
+
+     Config(
+          retries=2,
+          retry_handler=example_retry_handler
+          )
+
+
+``example_retry_handler`` should be a function defined by the user that will
+compute the retry cost for a particular failure, given some information about
+the failure.
+
+For example, the following handler will give a cost of 1 to all exceptions,
+except when a bash app exits with unix exitcode 9, in which case the cost will
+be 100. This will have the effect that retries will happen as normal for most
+errors, but the bash app can indicate that there is little point in retrying
+by exiting with exitcode 9.
+
```

Comment:
We should leave a note here that if the `retry_handler` itself fails, the cost is 1. Just to clarify that failure in the handler is treated similar to a failure in the app but with a pre-determined cost, just so that user's would avoid setting bad `retries` counts.

Reply:
If the retry_handler itself fails, the whole task is aborted so it isn't really correct to say the cost is 1. Really, the cost is infinity - where infinity means "something big enough to stop any more tries happening".
```
  task_record['fail_cost'] = self._config.retries + 1
```

I'll add in a note that says something like "if the retry handler fails, no further tries are attempted".

Reply:
@yadu I've added a documentation note for this in 313ef3b9c2bd2f49aebb3328803b454019ebb33b - can you check, and reapprove if ok?

Reply:
Alright this makes more sense.